### PR TITLE
feat: partial update/cache implementation

### PR DESCRIPTION
**Pull request by oscanezing**
Enable partial update in datarepo-go library

## JIRA Ticket URL
https://merlinapp.atlassian.net/browse/CM-234

## Description
Enable partial update in datarepo-go library

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I did add reviewers to this PR
- [ ]  My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I created the unit tests and they are passing

### CODE REVIEW

File: cached_repository.go [hunk 1]
```diff
@@ -16,6 +16,9 @@ type CachedRepository interface {
 	// If successful, then the value is also updated/cached in the configured caches and/or evicted
 	// from the caches that requested eviction on write operations
 	Update(ctx context.Context, value interface{}) error
+	// If successful, then the value is also updated/cached in the configured caches and/or evicted
```
Comment (camilomerlin) [inline cached_repository.go:21]: Can you update the comment to indicate that the fully updated object will be retrieved from the datastore and cached please?

File: data_writer.go [hunk 2]
```diff
@@ -9,4 +9,6 @@ type DataWriter interface {
 	Create(ctx context.Context, value interface{}) error
 	// Updates the provided value in the repository
 	Update(ctx context.Context, value interface{}) error
+	// Updates the provided value in the repository partially changing only received fields
+	PartialUpdate(ctx context.Context, value interface{}) error
```
Comment (camilomerlin) [inline data_writer.go:13]: 👍

File: integration_tests/book_gorm_redis/gorm_redis_int_nonuniquekey_test.go [hunk 3]
```diff
@@ -217,6 +217,43 @@ func (s *GormRedisIntegrationNonUniqueKeyTestSuite) TestGetUpdateAndGetTwoAuthor
...
+			author1 := testdomain.CreateAuthor(s.system)
+			author2 := testdomain.CreateAuthor(s.system)
+			book1, _ := author1.CreateBook(ctx, EmptyStatus)
+			book2, _ := author1.CreateBook(ctx, CompletedStatus)
+
+			authorIds := []string{author1.AuthorId, author2.AuthorId}
+
+			testdomain.GetBooksForAuthors(s.system, authorIds)
+
+			Convey("When I update the 1st book for the 1st author and query for the books again", func() {
```
Comment (camilomerlin) [inline integration_tests/book_gorm_redis/gorm_redis_int_nonuniquekey_test.go]: minor - `When I do a partial update of...`

File: integration_tests/book_gorm_redis/gorm_redis_int_nonuniquekey_test.go [hunk 4]
```diff
@@ -217,6 +217,43 @@ func (s *GormRedisIntegrationNonUniqueKeyTestSuite) TestGetUpdateAndGetTwoAuthor
 	})
 }
 
+func (s *GormRedisIntegrationNonUniqueKeyTestSuite) TestGetPartialUpdateAndGetTwoAuthorsBooks() {
```
Comment (camilomerlin) [inline integration_tests/book_gorm_redis/gorm_redis_int_nonuniquekey_test.go:220]: Nice! :)

File: cached_repository.go (hunk 1 above)
Comment (camilomerlin) [reply cached_repository.go:21]: Also, point out that there is a side-effect that the `value` is/can be used as an input/output, is that right?

The actual value changes after the operation in Gorm (assuming people use the Gorm backend). Please correct me if I'm wrong :)

**Review by pcu4dros — APPROVED** [review]

**Review by camilomerlin — APPROVED** [review]

File: integration_tests/book_gorm_redis/gorm_redis_int_nonuniquekey_test.go (hunk 3 above)
Comment (oscanezing) [reply integration_tests/book_gorm_redis/gorm_redis_int_nonuniquekey_test.go]: done :)

File: cached_repository.go (hunk 1 above)
Comment (oscanezing) [reply cached_repository.go:21]: yeah, received object is complemented with information that is saved in data storage, and `value` received is changed after gorm operation

Comment (oscanezing) [reply cached_repository.go:21]: done

**Pull request merged**